Add COS Lite integration with Charmed K8s (#838)
charmed-kubernetes/kubernetes-docs: pages/k8s/how-to-cos-lite.md

---
wrapper_template: "templates/docs/markdown.html"
markdown_includes:
  nav: "kubernetes/docs/shared/_side-navigation.md"
context:
  title: "Integrating COS Lite with Charmed Kubernetes"
  description: Integrating COS Lite with Charmed Kubernetes
keywords: Cloud, cluster, observability
tags: [observability, operating]
sidebar: k8smain-sidebar
permalink: how-to-cos-lite.html
layout: [base, ubuntu-com]
toc: False
---

First, initialize a controller on your preferred cloud platform (vSphere, AWS,
Azure) using Juju. Run the following command, ensuring to replace placeholder
values with your specifics:

```
juju bootstrap --credential $XYZ \
  --debug \
  --model-default automatically-retry-hooks=false \
  --model-default 'logging-config=<root>=DEBUG' \
  --bootstrap-image="juju-ci-root/templates/jammy-test-template" \
  --bootstrap-series=jammy \
  --bootstrap-constraints "arch=amd64" \
  aws tutorial-controller
```

Next, create a microk8s model to act as a deployment cloud for cos-lite:

```
juju add-model --config logging-config='<root>=DEBUG' \
  microk8s-ubuntu aws
```

Deploy Ubuntu on the microk8s-ubuntu model with the following specifications:

```

> juju deploy ubuntu --series=focal --constraints="mem=16G cores=8 root-disk=50G"

Reviewer: I would suggest it is more specific than that, because it gets confusing when you are referring to `microk8s`  itself. Could we call it something like 'microk8s-node' or 'o11y-k8s` or something less confusing

```

> To deploy microk8s on ubuntu, access the unit ubuntu using `juju ssh ubuntu/0`

Reviewer: ```suggestion
To deploy MicroK8s on Ubuntu, access the unit `ubuntu` using `juju ssh ubuntu/0`
```
remember to change the name ^ if altered above

and follow the configuration steps available at: <https://charmhub.io/topics/canonical-observability-stack/tutorials/install-microk8s#heading--configure-microk8s>

> After configuring microk8s, export its kubeconfig file to your current directory:

Reviewer: ```suggestion
After configuring MicroK8s, export its kubeconfig file to your current directory:
```

```
juju ssh ubuntu/0 -- microk8s config > microk8s-config.yaml
```

> Now, register microk8s as a Juju cloud:

Reviewer: ```suggestion
Now, register MicroK8s as a Juju cloud:
```

```
KUBECONFIG=microk8s-config.yaml juju add-k8s microk8s
```

Create a new model for cos-lite on this cloud and deploy cos-lite:

```
juju add-model cos-lite microk8s
juju deploy cos-lite
```

> Offer cos-lite's endpoints for integration across models:

Reviewer: isn't it 'COS-Lite'?

```
juju offer grafana:grafana-dashboard
juju offer prometheus:receive-remote-write
```

> Check the status of these offerings with juju status --relations to see

Reviewer: ```suggestion
Check the status of these offered interfaces with `juju status --relations` to see
```

> both grafana and prometheus listed.

Reviewer: ```suggestion
both Grafana and Prometheus listed.
```

This process bootstrapped a controller, created two models (one for Ubuntu

> with microk8s deployed, and another for cos-lite), and set up cos-lite's

Reviewer: ```suggestion
with MicroK8s deployed, and another for COS Lite), and set up COS Lite's
```

endpoints for cross-model integration.

> Next, proceed to deploy charmed Kubernetes and link it with cos-lite.

Reviewer: ```suggestion
Next, proceed to deploy Charmed Kubernetes and link it with COS Lite.
```

> Create a model for charmed kubernetes:

Reviewer: ```suggestion
Create a model for Charmed Kubernetes:
```

```
juju add-model --config logging-config='<root>=DEBUG' charmed-k8s aws
```

> Deploy charmed kubernetes:

Reviewer: ```suggestion
Deploy Charmed Kubernetes:
```

```
juju deploy charmed-kubernetes
```

> Consume cos-lite endpoints:

Reviewer: ```suggestion
Consume the COS Lite endpoints:
```

```
juju consume cos-lite.grafana
juju consume cos-lite.prometheus
```

> Deploy the grafana-agent:

Reviewer: ```suggestion
Deploy the `grafana-agent` charm:
```

```
Juju deploy grafana-agent
```

> Relate grafana-agent to k8s and kubernetes-control-plane and kubernetes-worker:

Reviewer: ```suggestion
Relate the `grafana-agent` charm to `k8s` and `kubernetes-control-plane` and `kubernetes-worker`:
```

```
juju integrate grafana-agent:cos-agent k8s:cos-agent
juju integrate grafana-agent:cos-agent kubernetes-control-plane:cos-agent
juju integrate grafana-agent:cos-agent kubernetes-worker:cos-agent
```

> Relate grafana-agent to the cos-lite offerings:

Reviewer: ```suggestion
Relate grafana-agent to the COS Lite offered interfaces:
```

```
juju integrate grafana-agent grafana
juju integrate grafana-agent prometheus
```

> Get the credentials and login url for grafana:

Reviewer: ```suggestion
Get the credentials and login URL for Grafana:
```

```
juju run grafana/0 get-admin-password -m cos-lite

> 

Reviewer: Please split the input from the output - a different code block for each at least, if not a line of explanatory text like "This will output the admin password and URL for Grafana:"

admin-password: b9OhxF5ndUDO
url: http://10.246.154.87/cos-lite-grafana

> ```

Reviewer 2: Definitely need a link to existing cos documents here to be clear how to use your shiny new observability stack

Reviewer: Yes. There may be something in tutorials or howtos, or a selection of things. It's fine to list several as a list. Maybe use a 'Next steps' heading.

---
wrapper_template: "templates/docs/markdown.html"
markdown_includes:
  nav: "kubernetes/docs/shared/_side-navigation.md"
context:
  title: "Integrating COS Lite with Charmed Kubernetes"
  description: Integrating COS Lite with Charmed Kubernetes
keywords: Cloud, cluster, observability
tags: [observability, operating]
sidebar: k8smain-sidebar
permalink: how-to-cos-lite.html
layout: [base, ubuntu-com]
toc: False
---

**Charmed Kubernetes** includes the standard Kubernetes dashboard for
monitoring your cluster. However, it is often advisable to have a monitoring
solution which will run whether the cluster itself is running or not. It may
also be useful to integrate monitoring into existing setups.

To make monitoring your cluster a delightful experience, Canonical provides

> first-class integration between charmed-kubernetes and COS Lite. This guide

Reviewer 3: Let's add this info somewhere please: COS stands for Canonical Observability Stack

> first-class integration between charmed-kubernetes and COS Lite. This guide

Reviewer: ```suggestion
first-class integration between Charmed Kubernetes and COS Lite. This guide
```

will help you integrate a COS Lite deployment with a charmed-kubernetes deployment.

This document assumes you have a controller with an installation of

> charmed-kubernetes. If this is not your case, refer to

Reviewer: ```suggestion
Charmed Kubernetes. If this is not your case, refer to
```

> ["how-to-install"](<https://ubuntu.com/kubernetes/docs/how-to-install>).

Reviewer: Put the links at the bottom of the page please. Also, this should use a relative link, in this case '/kubernetes/docs/how-to-install'

## Preparing a platform for COS Lite

> First, create a microk8s model to act as a deployment cloud for COS Lite:

Reviewer: ```suggestion
First, create a MicroK8s model to act as a deployment cloud for COS Lite:
```

```

> juju add-model --config logging-config='<root>=DEBUG' \

Reviewer 3: We can maybe add a little side note on the command: a juju model is similar to a kubernetes namespace and we are setting the logging config to debug so that you can look at logs using juju debug-log.

  microk8s-ubuntu aws
```

> Deploy the Ubuntu charm as an application called microk8s:

Reviewer: ```suggestion
Deploy the Ubuntu charm as an application called "microk8s":
```

```

> juju deploy ubuntu microk8s --series=focal --constraints="mem=16G cores=8 root-disk=50G"

Reviewer 3: Are these the minimal resource constraints?

Reviewer: 8 cores seems a bit hefty :thinking:

```

> To actually deploy microk8s on Ubuntu, access the unit ubuntu using `juju ssh

Reviewer: MicroK8s

microk8s/0` and follow the configuration steps available at: [Install

> Microk8s](<https://charmhub.io/topics/canonical-observability-stack/tutorials/install-microk8s#heading--configure-microk8s>)

Reviewer: Links at the end

After configuring microk8s, export its kubeconfig file to your current directory:

```
juju ssh microk8s/0 -- microk8s config > microk8s-config.yaml
```

> Now, register microk8s as a Juju cloud (see "[juju

Reviewer: MicroK8s

> add-k8s"](https://juju.is/docs/juju/juju-add-k8s) for details on the add-k8s

Reviewer: Link at the end

command):

```
KUBECONFIG=microk8s-config.yaml juju add-k8s microk8s
```

## Deploying COS Lite on the microk8s cloud

Create a new model for cos-lite on the microk8s cloud and deploy the cos-lite charm:

```

> juju add-model cos-lite microk8s

Reviewer 3: It doesn't have to be a part of this PR, maybe we can create a card to create a diagram that visualizes what we are doing?

Author: I think that's a good idea. I'll create a card.

juju deploy cos-lite
```

> Offer cos-lite's endpoints for integration across models:

Reviewer 3: Link somewhere to the juju cross-model relations please

```
juju offer grafana:grafana-dashboard
juju offer prometheus:receive-remote-write
```

Check the status of these offerings with `juju status --relations` to see
both grafana and prometheus listed.

So far, we've created a model that runs microk8s on Ubuntu, and added that
model as a Kubernetes cloud to Juju. We then used this cloud as a substrate
for the COS Lite deployment. We therefore have 2 models on the same controller.

This process created two models (one for Ubuntu with microk8s deployed, and
another for COS Lite), and set up COS Lite's endpoints for cross-model
integration. Next, proceed to integrate charmed-kubernetes with COS Lite.

## Integrating COS Lite with Charmed Kubernetes

Switch to your charmed-kubernetes model:

> `juju switch <charmed-kubernetes-model>`

Reviewer 3: Could you please add  something like: replace  <charmed-kubernetes-model> with your kubernetes model name, if you forgot the name run `juju models`

Consume the COS Lite endpoints:

```
juju consume cos-lite.grafana
juju consume cos-lite.prometheus
```

Deploy the grafana-agent:

```
Juju deploy grafana-agent
```

Relate grafana-agent to k8s, kubernetes-control-plane and kubernetes-worker:

```
juju integrate grafana-agent:cos-agent k8s:cos-agent
juju integrate grafana-agent:cos-agent kubernetes-control-plane:cos-agent
juju integrate grafana-agent:cos-agent kubernetes-worker:cos-agent
```

Relate grafana-agent to the cos-lite offerings:

```
juju integrate grafana-agent grafana
juju integrate grafana-agent prometheus
```

Get the credentials and login url for grafana:

```
juju run grafana/0 get-admin-password -m cos-lite

admin-password: b9OhxF5ndUDO
url: http://10.246.154.87/cos-lite-grafana
```

The username for this credential is `admin`.

Congratulations! You now have access to a complete observability stack when you
visit the url and enter the credentials.

Once you feel ready to dive deeper into your shiny new observability platform,

> you can head over to the [COS Lite documentation](https://charmhub.io/topics/canonical-observability-stack).

Reviewer: ```suggestion
you can head over to the [COS Lite documentation][].
```

> 

Reviewer: ```suggestion
<!-- LINKS -->
[COS Lite documentation]: https://charmhub.io/topics/canonical-observability-stack
```
^like this
